File: coding/3sum/starter.js
const threeSum = nums => {
  if (nums === null || nums.length < 3) return [];
  const triplets = [], n = nums.length;
  nums.sort((a,b) => a-b);
  for (let i = 0; i < n-2; i++){
      if (i > 0 && nums[i] === nums[i-1]) continue;
      let j = i + 1, k = n-1;
      while (j < k){
          let sum = nums[i] + nums[j] + nums[k];
          if (sum === 0){
              triplets.push([nums[i], nums[j], nums[k]]);
              j++;
              while (nums[j] === nums[j-1]) j++;
          } else if (sum < 0){
              j++;
              while (nums[j] === nums[j-1]) j++;
          } else {
              k--;
              while (nums[k] === nums[k+1]) k--;
          }
      }
  }

  return triplets;
}

File: coding/3sum/starter.py

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if not nums:
            return []
        triplets = []
        n = len(nums)
        nums.sort()
        for i in range(n-2):
            if i > 0 and nums[i] == nums[i-1]: continue
            j, k = i + 1, n - 1
            while j < k:
                sum = nums[i] + nums[j] + nums[k]
                if sum == 0:
                    triplets.append([nums[i], nums[j], nums[k]])
                    j += 1
                    k -= 1
                    while j < k and nums[j] == nums[j-1]: j += 1
                elif sum < 0:
                    j += 1
                else:
                    k -= 1

        return triplets
